File: src/advent_of_code_2022/advent_11.rs
use std::{fs, io};

pub fn solve() -> Result<(), io::Error> {
    let input = fs::read_to_string("data/advent_of_code_2022/input_11.txt").unwrap();
    let result_part_1 = solve_x(&input, true, 20);
    println!(
        "The level of monkey business after 20 rounds of stuff-slinging simian shenanigans is {}",
        result_part_1,
    );
    let result_part_2 = solve_x(&input, false, 10000);
    println!(
        "The level of monkey business after 10000 rounds is {}",
        result_part_2,
    );
    Ok(())
}

fn solve_x(input: &str, with_worry_update: bool, nr_round: i32) -> u64 {
    let monkey_strs: Vec<&str> = input.split("\n\n").collect();
    let mut monkeys: Vec<Monkey> = monkey_strs.into_iter().map(parse_monkey).collect();
    let common_divider: u64 = monkeys.iter().map(|monkey| monkey.divisible_by).product();

    for _n in 0..nr_round {
        for m in 0..monkeys.len() {
            let monkey = &mut monkeys[m];
            let mut items_to_throw: Vec<(usize, u64)> = vec![];
            while monkey.items.len() > 0 {
                let (target_monkey, item) = execute_round(&with_worry_update, monkey);
                items_to_throw.push((target_monkey, item % common_divider));
            }
            for (target_monkey, item) in items_to_throw {
                monkeys[target_monkey].items.push(item);
            }
        }
    }
    calculate_monkey_business(&monkeys)
}

fn execute_round(with_worry_update: &bool, monkey: &mut Monkey) -> (usize, u64) {
    let item = monkey.items.remove(0);
    let mut item_updated = match monkey.operation {
        Operation::MultiplyBy(value) => item * value,
        Operation::MultiplyByOld => item * item,
        Operation::Add(value) => item + value,
    };
    if *with_worry_update {
        item_updated = item_updated / 3;
    }
    monkey.nr_inspects += 1;
    let target_monkey = match item_updated % monkey.divisible_by {
        0 => monkey.if_true_throw_to_monkey,
        _ => monkey.if_false_throw_to_monkey,
    };
    (target_monkey, item_updated)
}

fn calculate_monkey_business(monkeys: &Vec<Monkey>) -> u64 {
    let mut result: Vec<u64> = monkeys.iter().map(|m| m.nr_inspects).collect();
    result.sort();
    result.reverse();
    result[0] * result[1]
}

fn parse_monkey(input: &str) -> Monkey {
    let lines: Vec<&str> = input.lines().collect();
    let (_, items_str): (&str, &str) = (lines[1].split_once(":")).unwrap();
    let items: Vec<u64> = items_str
        .split(",")
        .map(|x| x.trim().parse().unwrap())
        .collect();
    let operation = parse_operation(lines[2]);
    let (_, divisible_str) = (lines[3].split_once("by")).unwrap();
    let divisible_by: u64 = divisible_str.trim().parse().unwrap();

    let true_str = &lines[4]["    If true: throw to monkey ".len()..];
    let false_str = &lines[5]["    If false: throw to monkey ".len()..];
    Monkey {
        items,
        operation,
        divisible_by,
        if_true_throw_to_monkey: true_str.parse().unwrap(),
        if_false_throw_to_monkey: false_str.parse().unwrap(),
        nr_inspects: 0,
    }
}

fn parse_operation(line: &str) -> Operation {
    let (_, op_str) = (line.split_once("old")).unwrap();

    if &op_str[3..] == "old" {
        return Operation::MultiplyByOld;
    }
    let value: u64 = op_str[3..].parse().unwrap();
    return if op_str.trim().starts_with("*") {
        Operation::MultiplyBy(value)
    } else if op_str.trim().starts_with("+") {
        Operation::Add(value)
    } else {
        Operation::Add(0) // should not happen
    };
}

#[derive(PartialEq, Debug, Clone)]
enum Operation {
    MultiplyBy(u64),
    MultiplyByOld,
    Add(u64),
}

#[derive(PartialEq, Debug, Clone)]
struct Monkey {
    items: Vec<u64>,
    operation: Operation,
    divisible_by: u64,
    if_true_throw_to_monkey: usize,
    if_false_throw_to_monkey: usize,
    nr_inspects: u64,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::advent_of_code_2022::advent_11::Operation::MultiplyBy;

    #[test]
    fn test_parse_monkey() {
        let input = "Monkey 0:\n  Starting items: 79, 98\n  Operation: new = old * 19\n  Test: divisible by 23\n    If true: throw to monkey 2\n    If false: throw to monkey 3\n";
        let actual = parse_monkey(input);
        let expected = Monkey {
            items: vec![79, 98],
            operation: MultiplyBy(19),
            divisible_by: 23,
            if_true_throw_to_monkey: 2,
            if_false_throw_to_monkey: 3,
            nr_inspects: 0,
        };
        assert_eq!(actual, expected);
    }

    #[test]
    fn test_example_1() {
        let input = fs::read_to_string("data/advent_of_code_2022/example_11.txt").unwrap();
        let result = solve_x(&input, true, 20);
        assert_eq!(result, 10605);
    }

    #[test]
    fn test_example_2() {
        let input = fs::read_to_string("data/advent_of_code_2022/example_11.txt").unwrap();
        let result = solve_x(&input, false, 10000);
        assert_eq!(result, 2713310158);
    }
}

File: src/advent_of_code_2022/advent_11.py
import sys
from dataclasses import dataclass
from time import time
from typing import List


INPUT = """Monkey 0:
  Starting items: 57
  Operation: new = old * 13
  Test: divisible by 11
    If true: throw to monkey 3
    If false: throw to monkey 2

Monkey 1:
  Starting items: 58, 93, 88, 81, 72, 73, 65
  Operation: new = old + 2
  Test: divisible by 7
    If true: throw to monkey 6
    If false: throw to monkey 7

Monkey 2:
  Starting items: 65, 95
  Operation: new = old + 6
  Test: divisible by 13
    If true: throw to monkey 3
    If false: throw to monkey 5

Monkey 3:
  Starting items: 58, 80, 81, 83
  Operation: new = old * old
  Test: divisible by 5
    If true: throw to monkey 4
    If false: throw to monkey 5

Monkey 4:
  Starting items: 58, 89, 90, 96, 55
  Operation: new = old + 3
  Test: divisible by 3
    If true: throw to monkey 1
    If false: throw to monkey 7

Monkey 5:
  Starting items: 66, 73, 87, 58, 62, 67
  Operation: new = old * 7
  Test: divisible by 17
    If true: throw to monkey 4
    If false: throw to monkey 1

Monkey 6:
  Starting items: 85, 55, 89
  Operation: new = old + 4
  Test: divisible by 2
    If true: throw to monkey 2
    If false: throw to monkey 0

Monkey 7:
  Starting items: 73, 80, 54, 94, 90, 52, 69, 58
  Operation: new = old + 7
  Test: divisible by 19
    If true: throw to monkey 6
    If false: throw to monkey 0"""


def main():
    monkeys = parse_monkeys(INPUT)
    x = solve_x(monkeys)
    print(x)
    assert x == 28244037010


def solve_x(monkeys):
    common_divider = 1
    for d in [m.divisible_by for m in monkeys]:
        common_divider *= d

    for _n in range(10_000):
        for m in range(len(monkeys)):
            items_to_throw = []
            monkey = monkeys[m]
            while monkey.items:
                (target_monkey, item) = execute_round(monkey)
                items_to_throw.append((target_monkey, item % common_divider))

            for (target_monkey, item) in items_to_throw:
                monkeys[target_monkey].items.append(item)

    return calculate_monkey_business(monkeys)


def execute_round(monkey):
    item = monkey.items.pop(0)
    if monkey.operation == "multiply":
        item *= monkey.value
    elif monkey.operation == "add":
        item += monkey.value
    else:
        item *= item

    monkey.nr_inspects += 1
    if item % monkey.divisible_by == 0:
        target_monkey = monkey.if_true_throw_to_monkey
    else:
        target_monkey = monkey.if_false_throw_to_monkey

    return target_monkey, item


def calculate_monkey_business(monkeys):
    result = [m.nr_inspects for m in monkeys]
    result.sort()
    result.reverse()
    return result[0] * result[1]


@dataclass
class Monkey:
    items: List[int]
    operation: str
    value: int
    divisible_by: int
    if_true_throw_to_monkey: int
    if_false_throw_to_monkey: int
    nr_inspects: int = 0


def parse_operation(line):
    if line == '  Operation: new = old * old':
        return "square", 0
    op_str = line.split("old")[1]
    if op_str == "*":
        return "square", 0

    value = int(op_str[3:])

    if op_str.strip().startswith("*"):
        return "multiply", value
    if op_str.strip().startswith("+"):
        return "add", value


def parse_monkeys(s):
    result = []
    monkey_strs = s.split("\n\n")
    for monkey_str in monkey_strs:
        lines = monkey_str.splitlines()
        items_str = lines[1].split(":")[1]
        items = [int(s.strip()) for s in items_str.split(",")]
        (operation, value) = parse_operation(lines[2])
        divisible_by = int((lines[3].split("by")[1]).strip())
        a = len("    If true: throw to monkey ")
        b = len("    If false: throw to monkey ")

        if_true_throw_to_monkey = int(lines[4][a:])
        if_false_throw_to_monkey = int(lines[5][b:])
        result.append(Monkey(
            items=items,
            operation=operation,
            value=value,
            divisible_by=divisible_by,
            if_true_throw_to_monkey=if_true_throw_to_monkey,
            if_false_throw_to_monkey=if_false_throw_to_monkey,
            nr_inspects=0,
        ))
    return result


if __name__ == "__main__":
    t1 = time()
    main()
    t2 = time()
    print("Time elapsed: {} seconds".format(t2 - t1))
    sys.exit(0)
